cmd/juju/commands/bootstrap.go

```diff
@@ -126,7 +127,22 @@ agent stream is 'released' (default), then a 2.3.1 client can bootstrap:
     * 2.3.1 controller by running '... bootstrap ...';
     * 2.3.2 controller by running 'bootstrap --auto-upgrade'.
 However, if this client has a copy of codebase, then a local copy of Juju
-will be built and bootstrapped - 2.3.1.1.`
+will be built and bootstrapped - 2.3.1.1.
+
+If you are bootstrapping to a Kubernetes cluster that is not recognised by
+Juju as a cloud provider you may want to set the service type for the Juju
+controller. This will allow Juju clients to communicate with the Juju controller
+outside of the Kubernetes cluster. This can be controller using the
+controller-service-type --config option (see below).
+
+If Juju is bootstrapping to unknown Kubernetes cluster where a cloud can not be
+determined the following warning will be displayed during bootstrap
+"bootstrapping to an unknown kubernetes cluster should be used with option --config controller-service-type. See juju help bootstrap"
+
+Alternatively controller-service-type can be used to change the service type
+of the Juju controller in known clouds when the precomupted default is not
+adequate.
```

Comment:
Bootstrapping to a k8s cluster requires that the service set up to handle requests to the controller be accessible outside the cluster. Typically this means a service type of LoadBalancer is needed, and Juju does create such a service if it knows it is supported by the cluster.

When bootstrapping to a k8s cluster Juju does not recognise, there's no guarantee a load balancer is available, so Juju defaults to a controller service type of ClusterIP. This may not be suitable, so there's 3 bootstrap options available to tell Juju how to set up the controller service. Part of the solution may require a load balancer for the cluster to be set up manually first, or perhaps an external k8s service via a FQDN will be used (this is a cluster specific implementation decision which Juju needs to be informed about so it can set things up correctly). The 3 relevant bootstrap options are
- controller-service-type
- controller-external-name
- controller-external-ips

blah blah blah

Reply:
done & done